File: src/extractors/web_extractor.py
import logging
import time
import requests
from typing import List, Dict, Tuple, Optional, Callable
from urllib.parse import urlparse, urljoin
from bs4 import BeautifulSoup

from .markdown_converter import HTMLToStructuredMarkdown, _sanitize_text

logger = logging.getLogger(__name__)

# Rust optimizations if available
try:
    from content_extractor_rust_bridge import (
        decode_response_text,
        parse_image_dimension,
        extract_images as extract_images_rust,
        RUST_AVAILABLE,
    )
except ImportError:
    RUST_AVAILABLE = False
    decode_response_text = None
    parse_image_dimension = None
    extract_images_rust = None

_html_converter = HTMLToStructuredMarkdown()


def _extract_page_metadata(soup: BeautifulSoup, url: str) -> Dict:
    """Extract rich metadata from an HTML page."""
    meta = {"url": url}
    if soup.title and soup.title.string:
        meta["title"] = _sanitize_text(soup.title.string.strip())
    else:
        og = soup.find("meta", attrs={"property": "og:title"})
        meta["title"] = og["content"].strip() if og and og.get("content") else url

    desc = soup.find("meta", attrs={"name": "description"}) or soup.find("meta", attrs={"property": "og:description"})
    if desc and desc.get("content"):
        meta["description"] = _sanitize_text(desc["content"].strip()[:500])

    author = soup.find("meta", attrs={"name": "author"})
    if author and author.get("content"):
        meta["author"] = author["content"].strip()

    for attr in ["article:published_time", "datePublished", "date"]:
        tag = soup.find("meta", attrs={"property": attr}) or soup.find("meta", attrs={"name": attr})
        if tag and tag.get("content"):
            meta["date"] = tag["content"].strip()
            break

    html_tag = soup.find("html")
    if html_tag and html_tag.get("lang"):
        meta["language"] = html_tag["lang"]

    canonical = soup.find("link", attrs={"rel": "canonical"})
    if canonical and canonical.get("href"):
        meta["canonical"] = canonical["href"]

    return meta


class WebScanner:
    """World-class web crawler with structure-preserving extraction."""

    JUNK_IMG_PATTERNS = [
        "pixel",
        "tracking",
        "1x1",
        "spacer",
        "blank",
        ".gif",
        "data:image",
        "advertisement",
        "ad-",
        "analytics",
        "beacon",
        "logo",
        "favicon",
        "icon",
    ]

    def __init__(
        self,
        timeout: int = 15,
        user_agent: str = None,
        max_retries: int = 2,
        max_chars_per_page: int = 50_000,
    ):
        self.timeout = timeout
        self.user_agent = user_agent or (
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
            "(KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36 RAG_RAT/2.0"
        )
        self.max_retries = max_retries
        self.max_chars_per_page = max_chars_per_page
        self._seen_hashes = set()

    def scan(
        self,
        start_url: str,
        max_pages: int = 1,
        progress_callback: Optional[Callable] = None,
        page_callback: Optional[Callable[[str, Dict], None]] = None,
        completed_items: List[str] = None,
    ) -> Tuple[str, List[Dict], List[Dict]]:
        visited = set()
        queue = [start_url]
        base_domain = urlparse(start_url).netloc

        all_text = []
        all_images = []
        all_sources = []
        pages_crawled = 0

        while queue and pages_crawled < max_pages:
            url = queue.pop(0)
            if url in visited:
                continue
            visited.add(url)

            is_completed = completed_items and url in completed_items
            if progress_callback:
                status = f"Resuming: {url}" if is_completed else f"Scanning: {url}"
                try:
                    progress_callback(pages_crawled, max_pages, status, completed_item=url)
                except TypeError:
                    progress_callback(pages_crawled, max_pages)

            if is_completed:
                html = self._fetch_with_retry(url)
                if html is not None:
                    try:
                        soup = BeautifulSoup(html, "html.parser")
                        add_budget = max(0, max_pages - pages_crawled)
                        if add_budget > 0:
                            added = 0
                            for a in soup.find_all("a", href=True):
                                if added >= add_budget:
                                    break
                                link = self._normalize_link(a["href"], url, base_domain)
                                if link and link not in visited and link not in queue:
                                    queue.append(link)
                                    added += 1
                    except Exception as exc:
                        logger.debug("%s", exc)
                continue

            pages_crawled += 1
            html = self._fetch_with_retry(url)
            if html is None:
                continue

            try:
                soup = BeautifulSoup(html, "html.parser")
                meta = _extract_page_metadata(soup, url)
                page_images = self._extract_images(soup, url)
                all_images.extend(page_images)
                structured_text = _html_converter.convert(soup)
                structured_text = _sanitize_text(structured_text)
            except Exception as parse_err:
                logger.warning(f"Error parsing HTML from {url}: {parse_err}")
                continue

            content_hash = hash(structured_text[:2000])
            if content_hash in self._seen_hashes:
                continue
            self._seen_hashes.add(content_hash)

            if len(structured_text) > self.max_chars_per_page:
                structured_text = structured_text[: self.max_chars_per_page] + "\n\n[... truncated ...]"

            header = f"=== PAGE: {meta.get('title', url)} ===\nURL: {url}\n"
            if meta.get("description"):
                header += f"Description: {meta['description']}\n"
            header += "-" * 60 + "\n"

            page_block = header + structured_text
            all_text.append(page_block)

            source_entry = {
                "type": "web",
                "path": url,
                "title": meta.get("title", url),
                "chars": len(structured_text),
                "images": len(page_images),
            }
            all_sources.append(source_entry)

            if page_callback and not is_completed:
                try:
                    page_callback(page_block, source_entry)
                except Exception as exc:
                    logger.debug("%s", exc)

            if pages_crawled < max_pages:
                for a in soup.find_all("a", href=True):
                    link = self._normalize_link(a["href"], url, base_domain)
                    if link and link not in visited and link not in queue:
                        queue.append(link)

        return "\n\n".join(all_text), all_images[:500], all_sources

    def _fetch_with_retry(self, url: str) -> Optional[str]:
        for attempt in range(self.max_retries + 1):
            try:
                r = requests.get(url, timeout=self.timeout, headers={"User-Agent": self.user_agent})
                if r.status_code == 200:
                    if RUST_AVAILABLE and decode_response_text:
                        text = decode_response_text(r.content, r.encoding)
                        if text:
                            return text
                    return r.text
                elif r.status_code in (429, 503):
                    time.sleep(2 ** (attempt + 1))
                    continue
                return None
            except Exception as exc:
                logger.debug("%s", exc)
        return None

    def _extract_images(self, soup: BeautifulSoup, page_url: str) -> List[Dict]:
        if RUST_AVAILABLE and extract_images_rust:
            try:
                return extract_images_rust(str(soup), page_url)
            except Exception as exc:
                logger.debug("%s", exc)

        images = []
        seen_srcs = set()
        for img in soup.find_all("img"):
            src = img.get("src") or img.get("data-src") or img.get("data-lazy-src")
            if not src:
                continue
            src = urljoin(page_url, src)
            if any(x in src.lower() for x in self.JUNK_IMG_PATTERNS):
                continue
            if src in seen_srcs:
                continue
            seen_srcs.add(src)
            images.append({"url": src, "alt": img.get("alt", ""), "source": page_url})
        return images

    def _normalize_link(self, href: str, current_url: str, base_domain: str) -> Optional[str]:
        if not href or href.startswith(("#", "mailto:", "tel:", "javascript:")):
            return None
        href = urljoin(current_url, href)
        parsed = urlparse(href)
        if parsed.netloc != base_domain:
            return None
        return f"{parsed.scheme}://{parsed.netloc}{parsed.path}"

File: rust_output/src/web_extractor.rs
use anyhow::{Result, Context};
use crate::markdown_converter::{HTMLToStructuredMarkdown, _sanitize_text};
use std::collections::HashMap;
use std::collections::HashSet;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

pub static _HTML_CONVERTER: std::sync::LazyLock<HTMLToStructuredMarkdown> = std::sync::LazyLock::new(|| Default::default());

/// World-class web crawler with structure-preserving extraction.
#[derive(Debug, Clone)]
pub struct WebScanner {
    pub timeout: String,
    pub user_agent: String,
    pub max_retries: String,
    pub max_chars_per_page: String,
    pub _seen_hashes: HashSet<String>,
}

impl WebScanner {
    pub fn new(timeout: i64, user_agent: String, max_retries: i64, max_chars_per_page: i64) -> Self {
        Self {
            timeout,
            user_agent: (user_agent || "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36 RAG_RAT/2.0".to_string()),
            max_retries,
            max_chars_per_page,
            _seen_hashes: HashSet::new(),
        }
    }
    pub fn scan(&mut self, start_url: String, max_pages: i64, progress_callback: Option<Box<dyn Fn>>, page_callback: Option<Box<dyn Fn(serde_json::Value)>>, completed_items: Vec<String>) -> Result<(String, Vec<HashMap>, Vec<HashMap>)> {
        let mut visited = HashSet::new();
        let mut queue = vec![start_url];
        let mut base_domain = /* urlparse */ start_url.netloc;
        let mut all_text = vec![];
        let mut all_images = vec![];
        let mut all_sources = vec![];
        let mut pages_crawled = 0;
        while (queue && pages_crawled < max_pages) {
            let mut url = queue.remove(&0);
            if visited.contains(&url) {
                continue;
            }
            visited.insert(url);
            let mut is_completed = (completed_items && completed_items.contains(&url));
            if progress_callback {
                let mut status = if is_completed { format!("Resuming: {}", url) } else { format!("Scanning: {}", url) };
                // try:
                {
                    progress_callback(pages_crawled, max_pages, status, /* completed_item= */ url);
                }
                // except TypeError as _e:
            }
            if is_completed {
                let mut html = self._fetch_with_retry(url);
                if html.is_some() {
                    // try:
                    {
                        let mut soup = BeautifulSoup(html, "html.parser".to_string());
                        let mut add_budget = 0.max((max_pages - pages_crawled));
                        if add_budget > 0 {
                            let mut added = 0;
                            for a in soup.find_all("a".to_string(), /* href= */ true).iter() {
                                if added >= add_budget {
                                    break;
                                }
                                let mut link = self._normalize_link(a["href".to_string()], url, base_domain);
                                if (link && !visited.contains(&link) && !queue.contains(&link)) {
                                    queue.push(link);
                                    added += 1;
                                }
                            }
                        }
                    }
                    // except Exception as exc:
                }
                continue;
            }
            pages_crawled += 1;
            let mut html = self._fetch_with_retry(url);
            if html.is_none() {
                continue;
            }
            // try:
            {
                let mut soup = BeautifulSoup(html, "html.parser".to_string());
                let mut meta = _extract_page_metadata(soup, url);
                let mut page_images = self._extract_images(soup, url);
                all_images.extend(page_images);
                let mut structured_text = _html_converter.convert(soup);
                let mut structured_text = _sanitize_text(structured_text);
            }
            // except Exception as parse_err:
            let mut content_hash = hash(structured_text[..2000]);
            if self._seen_hashes.contains(&content_hash) {
                continue;
            }
            self._seen_hashes.insert(content_hash);
            if structured_text.len() > self.max_chars_per_page {
                let mut structured_text = (structured_text[..self.max_chars_per_page] + "\n\n[... truncated ...]".to_string());
            }
            let mut header = format!("=== PAGE: {} ===\nURL: {}\n", meta.get(&"title".to_string()).cloned().unwrap_or(url), url);
            if meta.get(&"description".to_string()).cloned() {
                header += format!("Description: {}\n", meta["description".to_string()]);
            }
            header += (("-".to_string() * 60) + "\n".to_string());
            let mut page_block = (header + structured_text);
            all_text.push(page_block);
            let mut source_entry = HashMap::from([("type".to_string(), "web".to_string()), ("path".to_string(), url), ("title".to_string(), meta.get(&"title".to_string()).cloned().unwrap_or(url)), ("chars".to_string(), structured_text.len()), ("images".to_string(), page_images.len())]);
            all_sources.push(source_entry);
            if (page_callback && !is_completed) {
                // try:
                {
                    page_callback(page_block, source_entry);
                }
                // except Exception as exc:
            }
            if pages_crawled < max_pages {
                for a in soup.find_all("a".to_string(), /* href= */ true).iter() {
                    let mut link = self._normalize_link(a["href".to_string()], url, base_domain);
                    if (link && !visited.contains(&link) && !queue.contains(&link)) {
                        queue.push(link);
                    }
                }
            }
        }
        Ok((all_text.join(&"\n\n".to_string()), all_images[..500], all_sources))
    }
    pub fn _fetch_with_retry(&mut self, url: String) -> Result<Option<String>> {
        for attempt in 0..(self.max_retries + 1).iter() {
            // try:
            {
                let mut r = /* reqwest::get( */&url).cloned().unwrap_or(/* timeout= */ self.timeout);
                if r.status_code == 200 {
                    if (RUST_AVAILABLE && decode_response_text) {
                        let mut text = decode_response_text(r.content, r.encoding);
                        if text {
                            text
                        }
                    }
                    r.text
                } else if (429, 503).contains(&r.status_code) {
                    std::thread::sleep(std::time::Duration::from_secs_f64((2)).pow((attempt + 1) as u32));
                    continue;
                }
                None
            }
            // except Exception as exc:
        }
        Ok(None)
    }
    pub fn _extract_images(&mut self, soup: BeautifulSoup, page_url: String) -> Result<Vec<HashMap>> {
        if (RUST_AVAILABLE && extract_images_rust) {
            // try:
            {
                extract_images_rust(soup.to_string(), page_url)
            }
            // except Exception as exc:
        }
        let mut images = vec![];
        let mut seen_srcs = HashSet::new();
        for img in soup.find_all("img".to_string()).iter() {
            let mut src = (img.get(&"src".to_string()).cloned() || img.get(&"data-src".to_string()).cloned() || img.get(&"data-lazy-src".to_string()).cloned());
            if !src {
                continue;
            }
            let mut src = urljoin(page_url, src);
            if self.JUNK_IMG_PATTERNS.iter().map(|x| src.to_lowercase().contains(&x)).collect::<Vec<_>>().iter().any(|v| *v) {
                continue;
            }
            if seen_srcs.contains(&src) {
                continue;
            }
            seen_srcs.insert(src);
            images.push(HashMap::from([("url".to_string(), src), ("alt".to_string(), img.get(&"alt".to_string()).cloned().unwrap_or("".to_string())), ("source".to_string(), page_url)]));
        }
        Ok(images)
    }
    pub fn _normalize_link(&self, href: String, current_url: String, base_domain: String) -> Option<String> {
        if (!href || href.starts_with(&*("#".to_string(), "mailto:".to_string(), "tel:".to_string(), "javascript:".to_string()))) {
            None
        }
        let mut href = urljoin(current_url, href);
        let mut parsed = /* urlparse */ href;
        if parsed.netloc != base_domain {
            None
        }
        format!("{}://{}{}", parsed.scheme, parsed.netloc, parsed.path)
    }
}

/// Extract rich metadata from an HTML page.
pub fn _extract_page_metadata(soup: BeautifulSoup, url: String) -> HashMap {
    // Extract rich metadata from an HTML page.
    let mut meta = HashMap::from([("url".to_string(), url)]);
    if (soup.title && soup.title.string) {
        meta["title".to_string()] = _sanitize_text(soup.title.string.trim().to_string());
    } else {
        let mut og = soup.find("meta".to_string(), /* attrs= */ HashMap::from([("property".to_string(), "og:title".to_string())]));
        meta["title".to_string()] = if (og && og.get(&"content".to_string()).cloned()) { og["content".to_string()].trim().to_string() } else { url };
    }
    let mut desc = (soup.find("meta".to_string(), /* attrs= */ HashMap::from([("name".to_string(), "description".to_string())])) || soup.find("meta".to_string(), /* attrs= */ HashMap::from([("property".to_string(), "og:description".to_string())])));
    if (desc && desc.get(&"content".to_string()).cloned()) {
        meta["description".to_string()] = _sanitize_text(desc["content".to_string()].trim().to_string()[..500]);
    }
    let mut author = soup.find("meta".to_string(), /* attrs= */ HashMap::from([("name".to_string(), "author".to_string())]));
    if (author && author.get(&"content".to_string()).cloned()) {
        meta["author".to_string()] = author["content".to_string()].trim().to_string();
    }
    for attr in vec!["article:published_time".to_string(), "datePublished".to_string(), "date".to_string()].iter() {
        let mut tag = (soup.find("meta".to_string(), /* attrs= */ HashMap::from([("property".to_string(), attr)])) || soup.find("meta".to_string(), /* attrs= */ HashMap::from([("name".to_string(), attr)])));
        if (tag && tag.get(&"content".to_string()).cloned()) {
            meta["date".to_string()] = tag["content".to_string()].trim().to_string();
            break;
        }
    }
    let mut html_tag = soup.find(&*"html".to_string()).map(|i| i as i64).unwrap_or(-1);
    if (html_tag && html_tag.get(&"lang".to_string()).cloned()) {
        meta["language".to_string()] = html_tag["lang".to_string()];
    }
    let mut canonical = soup.find("link".to_string(), /* attrs= */ HashMap::from([("rel".to_string(), "canonical".to_string())]));
    if (canonical && canonical.get(&"href".to_string()).cloned()) {
        meta["canonical".to_string()] = canonical["href".to_string()];
    }
    meta
}
